content/finding-command-help.md

```diff
@@ -0,0 +1,97 @@
+
+### help
+
+The `help` command displays all built-in commands supported by the current shell.
+
+**Use case**: A user wants to print basic commands supported by their shell into the current Terminal session.
+
+**Usage**: Type `help` into a Terminal session, then hit Enter.
+```bash
+help
+```
+![help output](/images/finding-command-help/help-output.png)
+
+### man
+
+`man` is short for manual pages. A command's manual page is a consistently structured text file describing the command's function, usage syntax, options, and similar commands. `man` opens a manpage in the terminal session, pressing <kbd>q</kbd> exits the manpage and returns the user to the shell prompt.
+
+**Use case**: A user wants to see actions performed by a specific command, and available options to modify the outcome of the command.
+
+**Usage**: Type `man` + any command into a Terminal session, then hit Enter. This example uses the `reboot` command.
+```bash
+man reboot
+```
+![man output](/images/finding-command-help/man-output.png)
+
+### info
+
+`info` pages are similar to man pages, however their descriptions for command usage are more in-depth and include examples of output. Users can also navigate to other related info pages using hypertext links within the info page. `info` opens an info page in the Terminal session, pressing <kbd>q</kbd> exits the manpage and returns the user to the shell prompt.
+
+**Use case**: A user wants to see examples of command usage, command output, and easily navigate to commands offering related or supporting functionality.
+
+**Usage**: Type `info` + any command into a Terminal session, then hit Enter. This example uses the `ls` command.
+```bash
+info ls
+```
+![info output](/images/finding-command-help/info-output.png)
+
+### apropos
+
+`apropos` performs a keyword search on command man pages. This is useful when searching for a command by its function. `apropos` prints output to the terminal session without navigating away from the shell prompt.
+
+**Use case**: A user can't remember the specific name of a command, but knows keywords commonly used to describe the command's functions.
+
+**Usage**: Type `apropos` + keyword. This example searches for commands related to the `delete` keyword.
+```bash
+apropos delete
+```
+![apropos output](/images/finding-command-help/apropos-output.png)
+
+### --help
+
+`--help` prints out a shortened version of the man page, usually including a brief example and all options for a command. `--help` prints output to the terminal session without navigating away from the shell prompt.
+
+**Use case**: A user wants to quickly reference all options for a known command in the current Terminal session.
+
+**Usage**: Type a command + `--help`. This example uses the `cd` command. Commands must be an exact match.
+```bash
+cd --help
+```
+![--help output](/images/finding-command-help/--help-output.png)
+
+### whatis
```

Comment:
Wondering if `which` + `ls -l` + `dpkg -S` to find out which package is providing a command would be useful to document here.

Along that line, information or tips for finding command info online may be helpful. (https://manpages.ubuntu.com/ or https://manpages.debian.org/ can be nicer to browse than doing everything in the terminal-- it's good to know the emergency options for when you're stuck in a TTY and have no internet or other device to read from, but that's a less common situation.)

Reply:
I didn't consider "which", but it is a good addition and will help set this article apart from the dozens of similar "finding help" resources that pop up in a Google search. I ran the command you mentioned "which ls -l dpkg -S" and noticed the output is the same as just "which ls dpkg". Is that expected?

Reply:
They were three separate commands. The use case goes like this: you know a command, and you want to know which package the command comes from (so you can install/uninstall it, look up documentation, etc.) First step, get path of the command:

```
jacob@serw12:~$ which libreoffice 
/usr/bin/libreoffice
```

Second, put that path into dpkg to search for what package provides it:

```
jacob@serw12:~$ dpkg -S /usr/bin/libreoffice 
libreoffice-common: /usr/bin/libreoffice
```

The issue with that is, dpkg only knows where the package wanted to install itself to, so if it's been symlinked somewhere else on your system, you may need to see where it's linking from and search for the original path instead of the one you're using:

```
jacob@serw12:~$ which shutdown
/usr/sbin/shutdown
jacob@serw12:~$ dpkg -S /usr/sbin/shutdown
dpkg-query: no path found matching pattern /usr/sbin/shutdown
jacob@serw12:~$ ls -alh /usr/sbin/shutdown
lrwxrwxrwx 1 root root 14 Jan 19 12:17 /usr/sbin/shutdown -> /bin/systemctl
jacob@serw12:~$ dpkg -S /bin/systemctl
systemd: /bin/systemctl
```

Unfortunately, even that sometimes doesn't work, because if a package requests to install itself into Dir A and your system has Dir A symlinked to Dir B, `which` and `ls` will both return Dir B, but `dpkg` will only know about Dir A. In those cases, you just have to know the common symlinks, like /bin to /usr/bin, /lib to /usr/lib, etc.

```
jacob@serw12:~$ which mkdir
/usr/bin/mkdir
jacob@serw12:~$ dpkg -S /usr/bin/mkdir
dpkg-query: no path found matching pattern /usr/bin/mkdir
jacob@serw12:~$ ls -alh /usr/bin/mkdir
-rwxr-xr-x 1 root root 91K Dec  5  2020 /usr/bin/mkdir
jacob@serw12:~$ ls -alh /bin
lrwxrwxrwx 1 root root 7 Oct 22  2020 /bin -> usr/bin
jacob@serw12:~$ dpkg -S /bin/mkdir
coreutils: /bin/mkdir
```

Reply:
I don't know how much of this needs to be included in this article, I just know looking up the package that owns a command is something I've done a number of times (on Pop systems with `dpkg -S` and with other package managers on other systems), and that's probably my top use case for `which` as well.

Reply:
Thanks for clarifying. Definitely useful, although I need to familiarize myself with this method. I'll add some version of this and make note of the symlink considerations.

Reply:
While poking around with this I noticed `which libreoffice | xargs dpkg -S` returns the correct package name, however `which libreoffice | xargs ls -alh` shows the path we passed to `dpkg -S` in the prior command (/usr/bin/libreoffice) is actually symlinked to ../lib/libreoffice/program/soffice. Is the concern that the symlinked files will return the "no matching path found" error, the user will assume the incorrect path, or both? I took my best guess in the most recent draft.

Reply:
What's going on with `libreoffice` is the following:
- dpkg installs the actual executable, `soffice`, to `/usr/lib/libreoffice/program/soffice`.
- dpkg also installs a symlink to `/usr/bin/libreoffice`; whoever created the LibreOffice packaging knew that people would want to start it with the command `libreoffice`, so they included that symlink in the package.
- When you ask dpkg (using the `-S` search flag) what package owns the file `/usr/bin/libreoffice`, you're asking it what package owns the symlink; because that symlink came from the package, dpkg is able to return the package name, even though it's not the actual path of the executable.

The concern would be the `no path found matching pattern ...` error. I think your last sentence "This verification step does not account for additional potential scenarios where a package installs to symlinked directory." is acceptable (if you don't want to go into the details of [the `/usr` merge](https://www.freedesktop.org/wiki/Software/systemd/TheCaseForTheUsrMerge/).) However, I would choose a different example for the `xargs ls -alh` command (pick one that _doesn't_ work with `dpkg -S` but does work after following the `ls` trail, such as `shutdown` linking to systemd from my comment above).

Also, your last paragraph refers to an `openoffice` command where you were trying to say `libreoffice`.

Reply:
Thanks for clarifying. This process provides good info but introduces a lot of concepts (path environment variables, /user merge, symlinks, etc.). In the latest draft I've tried to include it without going too far down the rabbit hole.